add go module support, manage vendor directory using `go mod vendor`

liggitt opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
part of https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/2019-03-19-go-modules.md

This PR:
* removes an unused dependency on go-ini
* records our reliance on various build tools to inform go module resolution, using a dummy `tools.go` file which has a build tag (`// +build tools`) ensuring it doesn't get built, and which adds imports of other packages we depend on
* creates `hack/update-vendor.sh` to generate and sync dependencies between the root and staging `go.mod` files, and to rebuild the `vendor` directory
* creates `hack/pin-dependency.sh` to record a specific tag/sha for a dependency
* updates `hack/update-vendor-licenses.sh` to generate the vendor license file from the vendored module listing
* removes godep.json files
* removes forked godep tool (and subsequently unused dependencies `github.com/kr/text` and `github.com/kr/pretty`)
* updates client-go install documentation with details about module use

**Notes for reviewer**

The vendor changes in the `update vendor` commit are due to:
* files for all build tags being included, which pulls in [appengine-specific files](https://github.com/kubernetes/kubernetes/blob/master/vendor/golang.org/x/oauth2/client_appengine.go) via golang.org/x/oauth (xref https://github.com/golang/go/issues/25873)
* removed golang.org/x/net/internal/nettest, test code is not vendored
* client-go doc changes are split into individual commits for ease of reviewing, will squash before merge

**Which issue(s) this PR fixes**:
xref #65683 https://github.com/kubernetes/kubernetes/issues/68577

Before merge:
- [x] test/document workflows for modifying dependencies:
  - [x] adding/updating/pinning a new dependency - `hack/pin-dependency.sh`
- [x] verify LICENSES file includes all previously included licenses:
  - https://github.com/kubernetes/kubernetes/pull/74877#issuecomment-479025860
- [x] email to kubernetes-dev describing change - https://groups.google.com/d/msg/kubernetes-dev/ECejidBCELI/Scc5zgGlBgAJ
- [x] update publishing bot in https://github.com/kubernetes/publishing-bot/pull/169:
  - rewrite staging go.mod files instead of godep.json files
  - construct synthetic godep.json files from go.mod dependencies
  - example run published to https://github.com/kubernetes-nightly/client-go/blob/master/go.mod

Follow-ups:
- [x] @sttts: push bot live once automated nightly test run completes
- [x] email to kubernetes-dev indicating changes have merged - https://groups.google.com/d/msg/kubernetes-dev/ECejidBCELI/dWC4iLiLCAAJ
- [x] PR to update https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/godep.md - https://github.com/kubernetes/community/pull/3612
- restructure license aggregation: https://github.com/kubernetes/kubernetes/issues/76093
- Relocate Godeps/LICENSES file to vendor/LICENSES: https://github.com/kubernetes/kubernetes/issues/76092
- Update CI to call verify-vendor / verify-vendor-licenses instead of verify-godeps (needs to be branch-specific?), and remove verify-godeps/verify-godep-licenses shims: https://github.com/kubernetes/test-infra/issues/12045 https://github.com/kubernetes/kubernetes/issues/76094 

```release-note
`k8s.io/kubernetes` and published components (like `k8s.io/client-go` and `k8s.io/api`) now publish go module files containing dependency version information. See http://git.k8s.io/client-go/INSTALL.md#go-modules for details on consuming `k8s.io/client-go` using go modules.
```

liggitt committed:
Remove unused github.com/go-ini/ini dep

liggitt committed:
Update hack scripts to use go mod

liggitt commented:
cc @dims @cblecker @sttts 

since we started running into trouble with things like golang.org/x/net, I thought it was worth exploring what it would actually take to get a reproducible vendor dir using go modules

liggitt commented:
@sttts this now has populated go.mod files in the staging repos, and dropped the godeps.json files to start flushing out dependencies on those.

for the verify-publishing-bot script, do we just need to know the full set of staging repos we have? if so, running this in each staging dir and accumulating the results produces the staging modules we have:
> GO111MODULE=on go list -m -json all | jq -r '. | select(.Main != true) | select(.Dir | contains("staging/src/k8s.io")) | .Path'

it would also be good to start thinking about what we'd want to end up in the published repos for those dirs.

I'd envision:
1. start with the go.mod file
2. transform the "STAGING REPLACE" replace() section:
  * limit to modules we actually require
  * switch the relative path directory reference to a sha of the published module
  * possibly run `go mod tidy` (not sure if we can point at a local repo and have it pick up a committed but not-yet-pushed sha)

idealhack commented:
/cc

liggitt committed:
Record code-generator as a dependency of repos using code generators

liggitt commented:
/retest

liggitt committed:
Update client-go docs for modules

- remove dep (c0a827dad6acc5fdea09967411aeeb9a8731d58f)
- move godep to bottom (3308b07da50c9547bcbfa50297b9bb91a02e88a2)
- bump example versions (9704cd1347ee287d388aa8f2b0091d4fe09640bb)
- add go modules section (e37037f5ae7c862a4255465ef328b8a3b6db038a)
- update go get docs (cda29fd9329a29856e5e3b218250b57ce8cbcd8f)

cblecker commented on build/builddeps/builddeps.go:
```diff
@@ -0,0 +1,41 @@
+/*
```
per https://github.com/kubernetes/community/pull/3493 and https://github.com/golang/go/wiki/Modules#how-can-i-track-tool-dependencies-for-a-module, you can use a build tag to import these dependencies, but never build them. In the community repo, I just created a `hack/tools.go` with a package named `tools`, and a build tag of `// +build tools` and it worked like a charm.

liggitt replied:
the build tag is a nice tweak, will update

thockin replied:
I learned all about modules today and @cblecker beat me to this :)

thockin commented on vendor/BUILD:
```diff
@@ -202,7 +202,6 @@ filegroup(
         "//vendor/github.com/ghodss/yaml:all-srcs",
         "//vendor/github.com/globalsign/mgo/bson:all-srcs",
         "//vendor/github.com/globalsign/mgo/internal/json:all-srcs",
-        "//vendor/github.com/go-ini/ini:all-srcs",
```
Do we know why the existing tools did not remove this automatically?

liggitt replied:
no idea. one does not peer into the madness that is godeps.

dims replied:
LOL!

thockin commented on hack/lib/init.sh:
```diff
@@ -22,6 +22,8 @@ set -o pipefail
 # https://github.com/kubernetes/kubernetes/issues/52255
 unset CDPATH
 
+export GO111MODULE=auto
```
Comment why?

thockin commented on hack/make-rules/verify.sh:
```diff
@@ -45,12 +45,13 @@ if [[ ${EXCLUDE_TYPECHECK:-} =~ ^[yY]$ ]]; then
     "verify-godeps.sh"             # runs in separate godeps job
-    "verify-staging-godeps.sh"     # runs in separate godeps job
-    "verify-godep-licenses.sh"     # runs in separate godeps job
+    "verify-godeps-licenses.sh"    # runs in separate godeps job
```
why are we keeping this one?

liggitt replied:
CI jobs reference it explicitly, and I want CI signal on this PR

thockin commented on hack/update-bazel.sh:
```diff
@@ -25,9 +25,11 @@ export GOBIN="${KUBE_OUTPUT_BINPATH}"
-go install k8s.io/kubernetes/vendor/github.com/bazelbuild/bazel-gazelle/cmd/gazelle
-go install k8s.io/kubernetes/vendor/github.com/bazelbuild/buildtools/buildozer
-go install k8s.io/kubernetes/vendor/k8s.io/repo-infra/kazel
+pushd "${KUBE_ROOT}/vendor"
```
>/dev/null 2>&1

thockin commented on hack/update-bazel.sh:
```diff
@@ -25,9 +25,11 @@ export GOBIN="${KUBE_OUTPUT_BINPATH}"
+  go install ./github.com/bazelbuild/bazel-gazelle/cmd/gazelle
+  go install ./github.com/bazelbuild/buildtools/buildozer
+  go install ./k8s.io/repo-infra/kazel
+popd
```
>/dev/null 2>&1

thockin commented on hack/update-vendor-licenses.sh:
```diff
@@ -147,7 +147,7 @@ fi
 cd "${LICENSE_ROOT}"
 
-GODEPS_LICENSE_FILE="Godeps/LICENSES"
+VENDOR_LICENSE_FILE="Godeps/LICENSES"
```
Do we have a plan to rename Godeps/ to something else?

thockin commented on hack/update-vendor-licenses.sh:
```diff
@@ -163,8 +163,13 @@ echo "= LICENSE $(cat "${LICENSE_ROOT}/LICENSE" | md5sum | awk '{print $1}')"
+for PACKAGE in $(GO111MODULE=on go list -mod=vendor -m -json all | jq -r .Path | sort -f); do
+  if [[ ! -e "${DEPS_DIR}/${PACKAGE}" ]]; then
+    echo "$PACKAGE doesn't exist in vendor, skipping" > /dev/stderr
+    continue
```
I don't think we should continue.  Why is this not a hard failure?

thockin commented on hack/update-vendor.sh:
```diff
@@ -0,0 +1,107 @@
+
+# create a symlink in vendor directory pointing to the staging client. This
+# let other packages use the staging client as if it were vendored.
+for repo in $(ls staging/src/k8s.io); do
```
Do we still need this?  I hoped that modules would eliminate this need

thockin commented on hack/update-vendor.sh:
```diff
@@ -0,0 +1,107 @@
+cat vendor/modules.txt | awk '{if($1=="#") print $2 " " $0; else print}' | sort -k1,1 -s | sed 's/.*#/#/' > vendor/modules.txt.tmp
+mv vendor/modules.txt.tmp vendor/modules.txt
+
+# create a symlink in vendor directory pointing to the staging client. This
```
s/client/repos ?

thockin commented on hack/update-vendor.sh:
```diff
@@ -0,0 +1,107 @@
+for repo in $(ls staging/src/k8s.io); do
+  echo "module k8s.io/${repo}
+
+go 1.12
```
rather than hardcode this - should we be using something more common?  Or does this version have consequences later?

E.g. When we update to go 1.13, 1.4, 1.15 will this script need to be updated?

liggitt replied:
switched to `go mod init` commands to set these up

thockin commented on hack/update-vendor.sh:
```diff
@@ -0,0 +1,107 @@
+source "${KUBE_ROOT}/hack/lib/init.sh"
+
+kube::log::status "Ensuring prereqs"
+kube::util::ensure_single_dir_gopath
```
Do we still need these prereqs?

liggitt replied:
nope, dropped

thockin commented on hack/update-vendor.sh:
```diff
@@ -0,0 +1,107 @@
+  ln -s "../../staging/src/k8s.io/${repo}" "${KUBE_ROOT}/vendor/k8s.io/${repo}"
+done
+
+# Workaround broken symlink in docker repo because godep copies the link, but
```
Do we still need this?

liggitt replied:
nope, dropped

thockin commented on hack/update-vendor.sh:
```diff
@@ -0,0 +1,107 @@
+  # change module name
+  # write to staging module file
+  cat "${KUBE_ROOT}/go.mod" | 
+  grep -v "k8s.io/${repo} " |
```
```
 | sed \
   -e  <expr> \
   -e <expr>  \
   -e <expr>

thockin commented on hack/update-vendor.sh:
```diff
@@ -0,0 +1,107 @@
+    
+    # remove double empty lines, staging and replace directives
+    cat go.mod | 
+    sed '/BEGIN STAGING REPLACE/,/END STAGING REPLACE/d' |
```
same sed comment

thockin commented on staging/src/k8s.io/code-generator/builddeps.go:
```diff
@@ -0,0 +1,29 @@
+/*
```
same build tags tricks?

thockin commented on go.mod:
```diff
@@ -217,3 +217,219 @@ require (
 )
+
+replace (
+	bitbucket.org/bertimus9/systemstat => bitbucket.org/bertimus9/systemstat v0.0.0-20180207000608-0eeff89b0690
```
I don't get it - these seem to be a duplication of the require block above - what's this about?

liggitt replied:
`require` notes a dependency of k8s.io/kubernetes on the given module at the given version (but go is free to pick a newer version if there are conflicting dependencies anywhere in our dependency graph)

`replace` pins the given module to the given version.

you need both in the top-level module being built to ensure that no dependency silently changes the versions you're pulling in.

thockin replied:
that seems super pointless - why would we `require` a specific tag only to have go change that under us?  I can see how it might matter for version ranges, but it seems pointless for examct specs?

liggitt replied:
welcome to go module [minimum version selection](https://github.com/golang/go/wiki/Modules#version-selection):

> The minimal version selection algorithm is used to select the versions of all modules used in a build. For each module in a build, the version selected by minimal version selection is always the semantically highest of the versions explicitly listed by a require directive in the main module or one of its dependencies.

thockin commented on go.mod:
```diff
@@ -59,214 +57,234 @@ require (
-	github.com/Rican7/retry v0.0.0-20160712041035-272ad122d6e5
+	github.com/NYTimes/gziphandler v0.0.0-20170623195520-56545f4a5d46 // indirect
+	github.com/Nvveen/Gotty v0.0.0-20120604004816-cd527374f1e5 // indirect
+	github.com/PuerkitoBio/purell v1.1.0
```
What did you run to do this work?  Why are we blindly updating some deps?

liggitt replied:
most of the commits have associated commands (`go mod init`, `go mod tidy`, `hack/update-vendor.sh`, etc). before this exits WIP, each commit will have a reproducible command associated/documented.

I'm actually trying to figure out why `go mod tidy` is bumping this particular dependency right now...

liggitt replied:
turns out go aggressively rewrites your go.mod file require statements to reflect the version of the referenced that will actually be used for the build, based on `max(local require, transitive dependency require...)`

the local `replace` can still stomp and pin to a particular version (which we do), but in this case, one of our dependencies was requiring v1.1.0, so our local require statement got bumped up:

```
$ go mod graph | grep github.com/PuerkitoBio/purell
k8s.io/kubernetes github.com/PuerkitoBio/purell@v1.1.0
github.com/go-openapi/loads@v0.17.0 github.com/PuerkitoBio/purell@v1.1.0
github.com/go-openapi/spec@v0.17.2 github.com/PuerkitoBio/purell@v1.1.0
github.com/go-openapi/loads@v0.17.2 github.com/PuerkitoBio/purell@v1.1.0
github.com/go-openapi/spec@v0.17.0 github.com/PuerkitoBio/purell@v1.1.0
github.com/go-openapi/jsonreference@v0.19.0 github.com/PuerkitoBio/purell@v1.1.0
github.com/go-openapi/jsonreference@v0.17.0 github.com/PuerkitoBio/purell@v1.1.0
```

liggitt replied:
all the commands (including the initialization commands used to take master from current state to this PR) are now baked into `hack/update-vendor.sh`

thockin reviewed (commented):
Is this all going to be reproducible when the current state inevitably changes underneath you?  It would be nice if the commands you ran were documented...

I can't approve the LICENSES change as is.  I need to verify it separately.

liggitt commented:
> Is this all going to be reproducible when the current state inevitably changes underneath you? It would be nice if the commands you ran were documented...

yeah, still WIP. will clean up and document before review.

> I can't approve the LICENSES change as is. I need to verify it separately.

understood. the Godeps-based one is package-level. the modules-based one is module-level.

thockin commented:
re: LICENSES -- there seem to be a ton of entries missing from it.  As a starting point `vendor/cloud.google.com/go/compute/metadata` is no longer here, but `vendor/cloud.google.com/go` is.  This may be technically correct (the license is on `vendor/cloud.google.com/go`) but is a HUGE change and much harder to verify.

We'll have to figure this out.

thockin commented:
wrt LICENSES - doing it at a package level means I can 1:1 verify.  Doing it per module makes that significantly harder.  Someone was working on a revamped collector that might make this easier.  Let me see if I can find that..

thockin commented:
e.g. if we collected a directory tree of license info, it would be easier to verify that any given package was subject to a higher-scope license.  Maybe even a per-package symlink back to the module's license info.

liggitt committed:
Add kubernetes/kubernetes tools.go imports

liggitt committed:
Add staging tools.go imports

liggitt commented:
/retest

dims commented:
/lgtm

(proactive LGTM :)

liggitt committed:
Remove third_party/forked/godep

liggitt commented:
this is ready for review

/hold 
for KEP approval, publishing bot work, and license approval (and resolution of how we want to construct LICENSES for modules)

cblecker commented:
@liggitt I take it this ready for review now?

explicit hold, as a change this big probably needs a couple sets of eyes before merge
/hold

cblecker commented:
Heh.. apparently I just need to refresh a little quicker.

Also: what communication is needed for this? Probably our own contributor base, but also I’m guessing downstream developers/consumers of our code too?

We should probably add our communication strategy to the KEP, thinking about it.

sttts commented:
> downstream developers/consumers of our code too?

Not for the master branch change. But we need something more extensive when we release the first 1.15 beta probably.

liggitt committed:
generated: hack/update-vendor.sh

liggitt commented:
/retest

liggitt commented:
/retest

liggitt commented:
@thockin, I did the following to verify that all licenses previously included are still included:

Licenses from master:
```
git show master:Godeps/LICENSES | egrep '= vendor/.*' | grep -v "licensed under" | sort -uf
```
<details>
```
= vendor/bitbucket.org/bertimus9/systemstat/LICENSE 6f7ba8122a9b3146266dafa39c5b8ee3
= vendor/cloud.google.com/go/LICENSE a873c5645c184d51e0f9b34e1d7cf559
= vendor/github.com/armon/circbuf/LICENSE d2d77030c0183e3d1e66d26dc1f243be
= vendor/github.com/asaskevich/govalidator/LICENSE 9548240229052f3a5f5bdf14ac19bbe3
= vendor/github.com/aws/aws-sdk-go/LICENSE.txt 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/Azure/azure-sdk-for-go/LICENSE cce6fd055830ca30ff78fdf077e870d6
= vendor/github.com/Azure/go-ansiterm/LICENSE 6000442264015a23894024af9930539b
= vendor/github.com/Azure/go-autorest/LICENSE a250e5ac3848f2acadb5adcb9555c18b
= vendor/github.com/bazelbuild/bazel-gazelle/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/bazelbuild/buildtools/LICENSE adb52eb384caedba181cd51fbcdf4b99
= vendor/github.com/beorn7/perks/LICENSE 0d0738f37ee8dc0b5f88a32e83c60198
= vendor/github.com/blang/semver/LICENSE 5a3ade42a900439691ebc22013660cae
= vendor/github.com/cespare/prettybench/LICENSE.txt 673caf450638c643cfede79f39c33087
= vendor/github.com/chai2010/gettext-go/LICENSE 87ce3ee0376881b02e75d3d5be2a6ba6
= vendor/github.com/client9/misspell/LICENSE 387f6b7d6741c8a7f4f7e3c2bbdf97e4
= vendor/github.com/cloudflare/cfssl/LICENSE 9bd1e7022303d9bbc29fda142f3e4fd0
= vendor/github.com/clusterhq/flocker-go/LICENSE d8103d9796cd0e951379d0834edad066
= vendor/github.com/codedellemc/goscaleio/LICENSE d2794c0df5b907fdace235a619d80314
= vendor/github.com/container-storage-interface/spec/LICENSE e3fc50a88d0a364313df4b21ef20c29e
= vendor/github.com/containerd/console/LICENSE 86d3f3a95c324c9479bd8986968f4327
= vendor/github.com/containerd/containerd/LICENSE.code aadc30f9c14d876ded7bedc0afd2d3d7
= vendor/github.com/containernetworking/cni/LICENSE fa818a259cbed7ce8bc2a22d35a464fc
= vendor/github.com/coreos/bbolt/LICENSE 13b2a308eefa10d841e3bf2467dbe07a
= vendor/github.com/coreos/etcd/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/coreos/go-oidc/LICENSE d2794c0df5b907fdace235a619d80314
= vendor/github.com/coreos/go-semver/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/coreos/go-systemd/LICENSE 19cbd64715b51267a47bf3750cc6a8a5
= vendor/github.com/coreos/pkg/LICENSE d2794c0df5b907fdace235a619d80314
= vendor/github.com/coreos/rkt/LICENSE 136e4f49dbf29942c572a3a8f6e88a77
= vendor/github.com/cpuguy83/go-md2man/LICENSE.md 80794f9009df723bbc6fe19234c9f517
= vendor/github.com/cyphar/filepath-securejoin/LICENSE 8d322afab99e1998dbfcc712f94e824d
= vendor/github.com/d2g/dhcp4/LICENSE 0187683301a45e8ea393bb2ffd7889c8
= vendor/github.com/d2g/dhcp4client/LICENSE b278a92d2c1509760384428817710378
= vendor/github.com/daaku/go.zipexe/license 2285ef50c2d5083f5bb92af715e5a378
= vendor/github.com/davecgh/go-spew/LICENSE 818c0a1d81cfcfdb7ecd58db268bab7e
= vendor/github.com/daviddengcn/go-colortext/LICENSE a802db7d9a036fb71c12cf20966510dc
= vendor/github.com/dgrijalva/jwt-go/LICENSE 276f2f3ba3749d25f6a6f5fb852d462e
= vendor/github.com/docker/distribution/LICENSE d2794c0df5b907fdace235a619d80314
= vendor/github.com/docker/docker/LICENSE 9740d093a080530b5c5c6573df9af45a
= vendor/github.com/docker/go-connections/LICENSE 04424bc6f5a5be60691b9824d65c2ad8
= vendor/github.com/docker/go-units/LICENSE 04424bc6f5a5be60691b9824d65c2ad8
= vendor/github.com/docker/libnetwork/LICENSE d2794c0df5b907fdace235a619d80314
= vendor/github.com/docker/spdystream/LICENSE 246dc1c1661293602d98ff9113c3be48
= vendor/github.com/elazarl/goproxy/LICENSE e2e14e5f5bd856768da39707ed93cd41
= vendor/github.com/emicklei/go-restful/LICENSE 2ebc1c12a0f4eae5394522e31961e1de
= vendor/github.com/euank/go-kmsg-parser/LICENSE e3fc50a88d0a364313df4b21ef20c29e
= vendor/github.com/evanphx/json-patch/LICENSE 72c842ec53c3334a81b6d65b6f9025a3
= vendor/github.com/exponent-io/jsonpath/LICENSE 42f582355f11b1d4bc8615214b7f0c38
= vendor/github.com/fatih/camelcase/LICENSE.md 4c59b216ce25dc182cdb837e07ba07c0
= vendor/github.com/fsnotify/fsnotify/LICENSE c38914c9a7ab03bb2b96d4baaee10769
= vendor/github.com/GeertJohan/go.rice/LICENSE d5ffb28cde0ca932cbab7e1e1d52c20e
= vendor/github.com/ghodss/yaml/LICENSE 0ceb9ff3b27d3a8cf451ca3785d73c71
= vendor/github.com/globalsign/mgo/LICENSE 566e96676859b5704130b80941bc9f1f
= vendor/github.com/go-ini/ini/LICENSE 19cbd64715b51267a47bf3750cc6a8a5
= vendor/github.com/go-openapi/analysis/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/go-openapi/errors/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/go-openapi/jsonpointer/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/go-openapi/jsonreference/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/go-openapi/loads/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/go-openapi/runtime/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/go-openapi/spec/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/go-openapi/strfmt/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/go-openapi/swag/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/go-openapi/validate/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/go-ozzo/ozzo-validation/LICENSE da12d993f2ce14947ad6eec35520b081
= vendor/github.com/go-sql-driver/mysql/LICENSE 815ca599c9df247a0c7f619bab123dad
= vendor/github.com/godbus/dbus/LICENSE 09042bd5c6c96a2b9e45ddf1bc517eed
= vendor/github.com/gogo/protobuf/LICENSE f76ab0572aa28f537b1508f2f2bc155e
= vendor/github.com/golang/groupcache/LICENSE 19cbd64715b51267a47bf3750cc6a8a5
= vendor/github.com/golang/mock/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/golang/protobuf/LICENSE 14db3a56c3796a940ba32948a15f97d0
= vendor/github.com/google/btree/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/google/cadvisor/LICENSE e7790b946bfacb700e8a8f2baedb3205
= vendor/github.com/google/certificate-transparency-go/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/google/gofuzz/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/google/uuid/LICENSE 88073b6dd8ec00fe09da59e0b6dfded1
= vendor/github.com/googleapis/gnostic/LICENSE b1e01b26bacfc2232046c90a330332b3
= vendor/github.com/GoogleCloudPlatform/k8s-cloud-provider/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/gophercloud/gophercloud/LICENSE dd19699707373c2ca31531a659130416
= vendor/github.com/gorilla/websocket/LICENSE c007b54a1743d596f46b2748d9f8c044
= vendor/github.com/gregjones/httpcache/LICENSE.txt 3cfef421226b2dacde78a4871380ac24
= vendor/github.com/grpc-ecosystem/go-grpc-middleware/LICENSE 7ab5c73bb7e4679b16dd7c11b3559acf
= vendor/github.com/grpc-ecosystem/go-grpc-prometheus/LICENSE 7ab5c73bb7e4679b16dd7c11b3559acf
= vendor/github.com/grpc-ecosystem/grpc-gateway/LICENSE.txt c510a2a01572b82d27f28fd4d02ca318
= vendor/github.com/hashicorp/golang-lru/LICENSE f27a50d2e878867827842f2c60e30bfc
= vendor/github.com/hashicorp/hcl/LICENSE b278a92d2c1509760384428817710378
= vendor/github.com/heketi/heketi/LICENSE a58e72c3bda574189508cb90d56fa19f
= vendor/github.com/imdario/mergo/LICENSE ff13e03bb57bf9c52645f2f942afa28b
= vendor/github.com/inconshreveable/mousetrap/LICENSE b23cff9db13f093a4e6ff77105cbd8eb
= vendor/github.com/JeffAshton/win_pdh/LICENSE fadcabe0503181faf8d4a9579bed3b7f
= vendor/github.com/jmespath/go-jmespath/LICENSE 9abfa8353fce3f2cb28364e1e9016852
= vendor/github.com/jmhodges/clock/LICENSE bc91f6bbbc0460d728657b023f3f6596
= vendor/github.com/jmoiron/sqlx/LICENSE 4353e10ea313ec47121dd2c7d89ce6b9
= vendor/github.com/jonboulle/clockwork/LICENSE 136e4f49dbf29942c572a3a8f6e88a77
= vendor/github.com/json-iterator/go/LICENSE 0b69744b481d72d30dbf69f84a451cfd
= vendor/github.com/jstemmer/go-junit-report/LICENSE 5d54de929d5bc244d21d8d4d91a6ac64
= vendor/github.com/jteeuwen/go-bindata/LICENSE 8dcedca69f7a474372829521f37954b1
= vendor/github.com/kardianos/osext/LICENSE 591778525c869cdde0ab5a1bf283cd81
= vendor/github.com/karrick/godirwalk/LICENSE 7bea66fc0a31c6329f9392034bee75d2
= vendor/github.com/kisielk/sqlstruct/LICENSE bc5911f8d3b8034f564b63d7ea436519
= vendor/github.com/kr/fs/LICENSE 591778525c869cdde0ab5a1bf283cd81
= vendor/github.com/kr/pretty/License 9d305c2010c6891ee4f3cd42a562f78f
= vendor/github.com/kr/text/License 449bfedd81a372635934cf9ce004c0cf
= vendor/github.com/lib/pq/LICENSE.md 0eab29964025b358179aa6d8f7db14bf
= vendor/github.com/libopenstorage/openstorage/LICENSE 40c3e1c9eacda859a17048003909a2f8
= vendor/github.com/liggitt/tabwriter/LICENSE 5d4950ecb7b26d2c5e4e7b4e0dd74707
= vendor/github.com/lithammer/dedent/LICENSE 2f1370f16b018459285c6972a934ba0e
= vendor/github.com/magiconair/properties/LICENSE c383a608fb9a0d227953e928803b9631
= vendor/github.com/mailru/easyjson/LICENSE 819e81c2ec13e1bbc47dc5e90bb4d88b
= vendor/github.com/MakeNowJust/heredoc/LICENSE 59c4411f6d7dfdaa85623e672d3d4438
= vendor/github.com/marstr/guid/LICENSE.txt 356484d12e6ad8a7c2d360b236e9a9c8
= vendor/github.com/mattn/go-shellwords/LICENSE e5116fc268f5118168ff06a271b50ef9
= vendor/github.com/mattn/go-sqlite3/LICENSE 2b7590a6661bc1940f50329c495898c6
= vendor/github.com/matttproud/golang_protobuf_extensions/LICENSE e3fc50a88d0a364313df4b21ef20c29e
= vendor/github.com/mesos/mesos-go/LICENSE 6c4db32a2fa8717faffa1d4f10136f47
= vendor/github.com/mholt/caddy/LICENSE.txt e3fc50a88d0a364313df4b21ef20c29e
= vendor/github.com/Microsoft/go-winio/LICENSE 69205ff73858f2c22b2ca135b557e8ef
= vendor/github.com/Microsoft/hcsshim/LICENSE d4c2cbbea5ee1e7c86dff68a7073718e
= vendor/github.com/miekg/dns/LICENSE 147353de6868a20caa562d26eab7b3c5
= vendor/github.com/mindprince/gonvml/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/mistifyio/go-zfs/LICENSE cce9462224bfb44c1866ef7bd5eddf54
= vendor/github.com/mitchellh/go-wordwrap/LICENSE.md 56da355a12d4821cda57b8f23ec34bc4
= vendor/github.com/mitchellh/mapstructure/LICENSE 3f7765c3d4f58e1f84c4313cecf0f5bd
= vendor/github.com/modern-go/concurrent/LICENSE 86d3f3a95c324c9479bd8986968f4327
= vendor/github.com/modern-go/reflect2/LICENSE 86d3f3a95c324c9479bd8986968f4327
= vendor/github.com/mohae/deepcopy/LICENSE 268dc9c546e3de67a93c1d12a039d702
= vendor/github.com/mrunalp/fileutils/LICENSE 435b266b3899aa8a959f17d41c56def8
= vendor/github.com/munnerz/goautoneg/README.txt a33eda65f1bc658f358e1d690c6a93d4
= vendor/github.com/mvdan/xurls/LICENSE d459301ff3fc917904837249508d99af
= vendor/github.com/mxk/go-flowrate/LICENSE 781f3c14fa867bae781c9e409831f954
= vendor/github.com/Nvveen/Gotty/LICENSE ab11220c6af5f3758ccae7bf525f8dec
= vendor/github.com/NYTimes/gziphandler/LICENSE.md e30b94cbe70132b181f72f953fbb3c82
= vendor/github.com/onsi/ginkgo/LICENSE 570603114d52313cb86c0206401c9af7
= vendor/github.com/onsi/gomega/LICENSE 570603114d52313cb86c0206401c9af7
= vendor/github.com/opencontainers/go-digest/LICENSE.code 9cd86830b557232ce55e2a6b47387471
= vendor/github.com/opencontainers/image-spec/LICENSE 27ef03aa2da6e424307f102e8b42621d
= vendor/github.com/opencontainers/runc/LICENSE 435b266b3899aa8a959f17d41c56def8
= vendor/github.com/opencontainers/runtime-spec/LICENSE b355a61a394a504dacde901c958f662c
= vendor/github.com/opencontainers/selinux/LICENSE e3fc50a88d0a364313df4b21ef20c29e
= vendor/github.com/pborman/uuid/LICENSE 88073b6dd8ec00fe09da59e0b6dfded1
= vendor/github.com/pelletier/go-toml/LICENSE dc9ea87a81f62b8871b2a4158edbfde6
= vendor/github.com/peterbourgon/diskv/LICENSE f9f3e815fc84aa04c4f4db33c553eef9
= vendor/github.com/pkg/errors/LICENSE 6fe682a02df52c6653f33bd0f7126b5a
= vendor/github.com/pkg/sftp/LICENSE 452fc5cc5a9127a0e828d73423d45035
= vendor/github.com/pmezard/go-difflib/LICENSE e9a2ebb8de779a07500ddecca806145e
= vendor/github.com/pquerna/cachecontrol/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/pquerna/ffjson/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/prometheus/client_golang/LICENSE 86d3f3a95c324c9479bd8986968f4327
= vendor/github.com/prometheus/client_model/LICENSE 86d3f3a95c324c9479bd8986968f4327
= vendor/github.com/prometheus/common/LICENSE 86d3f3a95c324c9479bd8986968f4327
= vendor/github.com/prometheus/procfs/LICENSE 86d3f3a95c324c9479bd8986968f4327
= vendor/github.com/PuerkitoBio/purell/LICENSE fb8b39492731abb9a3d68575f3eedbfa
= vendor/github.com/PuerkitoBio/urlesc/LICENSE 591778525c869cdde0ab5a1bf283cd81
= vendor/github.com/quobyte/api/LICENSE beacc5ea3bcda24bdcec545022dbb0b4
= vendor/github.com/rancher/go-rancher/LICENSE 2ee41112a44fe7014dce33e26468ba93
= vendor/github.com/Rican7/retry/LICENSE 6421d9994f58f630cd4dd33e88991ccc
= vendor/github.com/robfig/cron/LICENSE d924a63cb54a2a6c4bd28c50b2b0af59
= vendor/github.com/rubiojr/go-vhd/LICENSE 9ce5db55ba47444787183e59733e1977
= vendor/github.com/russross/blackfriday/LICENSE.txt ecf8a8a60560c35a862a4a545f2db1b3
= vendor/github.com/satori/go.uuid/LICENSE ae4ba217c6e20c2d8f48f69966b9121b
= vendor/github.com/seccomp/libseccomp-golang/LICENSE 343b433e752e8b44a543cdf61f14b628
= vendor/github.com/shurcooL/sanitized_anchor_name/LICENSE bee2c3aa5bd0f265ffbd193eb18ca30d
= vendor/github.com/sigma/go-inotify/LICENSE 5d4950ecb7b26d2c5e4e7b4e0dd74707
= vendor/github.com/sirupsen/logrus/LICENSE 8dadfef729c08ec4e631c4f6fc5d43a0
= vendor/github.com/soheilhy/cmux/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/spf13/afero/LICENSE.txt 920d76114a32b0fb75b3f2718c5a91be
= vendor/github.com/spf13/cast/LICENSE 67fac7567cbf6ba946e5576d590b1ed4
= vendor/github.com/spf13/cobra/LICENSE.txt 920d76114a32b0fb75b3f2718c5a91be
= vendor/github.com/spf13/jwalterweatherman/LICENSE 67fac7567cbf6ba946e5576d590b1ed4
= vendor/github.com/spf13/pflag/LICENSE 1e8b7dc8b906737639131047a590f21d
= vendor/github.com/spf13/viper/LICENSE 67fac7567cbf6ba946e5576d590b1ed4
= vendor/github.com/storageos/go-api/LICENCE 10d8703157b5fd9422b1813bae555905
= vendor/github.com/stretchr/objx/LICENSE.md 21e79cfe7201b9b64535bfae0895795b
= vendor/github.com/stretchr/testify/LICENSE 39cd1d751bc25944831de86496e3cf68
= vendor/github.com/syndtr/gocapability/LICENSE a7304f5073e7be4ba7bffabbf9f2bbca
= vendor/github.com/tmc/grpc-websocket-proxy/LICENSE 0894f9b225c28f0896b4bab675a2e19a
= vendor/github.com/ugorji/go/LICENSE a03c0693c900925da5789db4e72da142
= vendor/github.com/vishvananda/netlink/LICENSE 2ade771c7d7211af507864e8dd520529
= vendor/github.com/vishvananda/netns/LICENSE 2ade771c7d7211af507864e8dd520529
= vendor/github.com/vmware/govmomi/LICENSE.txt 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/vmware/photon-controller-go-sdk/LICENSE 0de60303c844eac44e45012dac1987de
= vendor/github.com/xanzy/go-cloudstack/LICENSE 136e4f49dbf29942c572a3a8f6e88a77
= vendor/github.com/xiang90/probing/LICENSE 61b57bfd44f7924c8a12cd72d3aa1e93
= vendor/go.uber.org/atomic/LICENSE.txt 1caee86519456feda989f8a838102b50
= vendor/go.uber.org/multierr/LICENSE.txt f65b21a547112d1bc7b11b90f9b31997
= vendor/go.uber.org/zap/LICENSE.txt 5e8153e456a82529ea845e0d511abb69
= vendor/golang.org/x/crypto/LICENSE 5d4950ecb7b26d2c5e4e7b4e0dd74707
= vendor/golang.org/x/lint/LICENSE 4c728948788b1a02f33ae4e906546eef
= vendor/golang.org/x/net/LICENSE 5d4950ecb7b26d2c5e4e7b4e0dd74707
= vendor/golang.org/x/oauth2/LICENSE 704b1e0c436dbf193e7dcbd4cf06ec81
= vendor/golang.org/x/sys/LICENSE 5d4950ecb7b26d2c5e4e7b4e0dd74707
= vendor/golang.org/x/text/LICENSE 5d4950ecb7b26d2c5e4e7b4e0dd74707
= vendor/golang.org/x/time/LICENSE 5d4950ecb7b26d2c5e4e7b4e0dd74707
= vendor/golang.org/x/tools/LICENSE 5d4950ecb7b26d2c5e4e7b4e0dd74707
= vendor/gonum.org/v1/gonum/LICENSE 665e67d07d85e236cceb8de602c6255a
= vendor/google.golang.org/api/LICENSE a651bb3d8b1c412632e28823bb432b40
= vendor/google.golang.org/genproto/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/google.golang.org/grpc/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/gopkg.in/gcfg.v1/LICENSE 13cea479df204c85485b5db6eb1bc9d5
= vendor/gopkg.in/inf.v0/LICENSE 13cea479df204c85485b5db6eb1bc9d5
= vendor/gopkg.in/natefinch/lumberjack.v2/LICENSE 574cdb55b81249478f5af5f789e9e29f
= vendor/gopkg.in/square/go-jose.v2/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/gopkg.in/warnings.v0/LICENSE c6775875c9d604beb22447dfae3d7049
= vendor/gopkg.in/yaml.v2/LICENSE e3fc50a88d0a364313df4b21ef20c29e
= vendor/k8s.io/gengo/LICENSE ad09685d909e7a9f763d2bb62d4bd6fb
= vendor/k8s.io/heapster/LICENSE 136e4f49dbf29942c572a3a8f6e88a77
= vendor/k8s.io/klog/LICENSE 19cbd64715b51267a47bf3750cc6a8a5
= vendor/k8s.io/kube-openapi/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/k8s.io/repo-infra/LICENSE e3fc50a88d0a364313df4b21ef20c29e
= vendor/k8s.io/utils/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/sigs.k8s.io/kustomize/LICENSE e3fc50a88d0a364313df4b21ef20c29e
= vendor/sigs.k8s.io/structured-merge-diff/LICENSE e3fc50a88d0a364313df4b21ef20c29e
= vendor/sigs.k8s.io/yaml/LICENSE 0ceb9ff3b27d3a8cf451ca3785d73c71
= vendor/vbom.ml/util/LICENSE 9f7e1d7e8f527330ebb5f4c32e0f3e40
```
</details>

Licenses from this PR:
```
git show HEAD:Godeps/LICENSES | egrep '= vendor/.*' | grep -v "licensed under" | sort -uf > head.licenses.txt
```
<details>
```
= vendor/bitbucket.org/bertimus9/systemstat/LICENSE 6f7ba8122a9b3146266dafa39c5b8ee3
= vendor/cloud.google.com/go/LICENSE a873c5645c184d51e0f9b34e1d7cf559
= vendor/github.com/armon/circbuf/LICENSE d2d77030c0183e3d1e66d26dc1f243be
= vendor/github.com/asaskevich/govalidator/LICENSE 9548240229052f3a5f5bdf14ac19bbe3
= vendor/github.com/aws/aws-sdk-go/LICENSE.txt 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/Azure/azure-sdk-for-go/LICENSE cce6fd055830ca30ff78fdf077e870d6
= vendor/github.com/Azure/go-ansiterm/LICENSE 6000442264015a23894024af9930539b
= vendor/github.com/Azure/go-autorest/LICENSE a250e5ac3848f2acadb5adcb9555c18b
= vendor/github.com/bazelbuild/bazel-gazelle/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/bazelbuild/buildtools/LICENSE adb52eb384caedba181cd51fbcdf4b99
= vendor/github.com/beorn7/perks/LICENSE 0d0738f37ee8dc0b5f88a32e83c60198
= vendor/github.com/blang/semver/LICENSE 5a3ade42a900439691ebc22013660cae
= vendor/github.com/cespare/prettybench/LICENSE.txt 673caf450638c643cfede79f39c33087
= vendor/github.com/chai2010/gettext-go/LICENSE 87ce3ee0376881b02e75d3d5be2a6ba6
= vendor/github.com/client9/misspell/LICENSE 387f6b7d6741c8a7f4f7e3c2bbdf97e4
= vendor/github.com/cloudflare/cfssl/LICENSE 9bd1e7022303d9bbc29fda142f3e4fd0
= vendor/github.com/clusterhq/flocker-go/LICENSE d8103d9796cd0e951379d0834edad066
= vendor/github.com/codedellemc/goscaleio/LICENSE d2794c0df5b907fdace235a619d80314
= vendor/github.com/container-storage-interface/spec/LICENSE e3fc50a88d0a364313df4b21ef20c29e
= vendor/github.com/containerd/console/LICENSE 86d3f3a95c324c9479bd8986968f4327
= vendor/github.com/containerd/containerd/LICENSE.code aadc30f9c14d876ded7bedc0afd2d3d7
= vendor/github.com/containernetworking/cni/LICENSE fa818a259cbed7ce8bc2a22d35a464fc
= vendor/github.com/coreos/bbolt/LICENSE 13b2a308eefa10d841e3bf2467dbe07a
= vendor/github.com/coreos/etcd/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/coreos/go-oidc/LICENSE d2794c0df5b907fdace235a619d80314
= vendor/github.com/coreos/go-semver/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/coreos/go-systemd/LICENSE 19cbd64715b51267a47bf3750cc6a8a5
= vendor/github.com/coreos/pkg/LICENSE d2794c0df5b907fdace235a619d80314
= vendor/github.com/coreos/rkt/LICENSE 136e4f49dbf29942c572a3a8f6e88a77
= vendor/github.com/cpuguy83/go-md2man/LICENSE.md 80794f9009df723bbc6fe19234c9f517
= vendor/github.com/cyphar/filepath-securejoin/LICENSE 8d322afab99e1998dbfcc712f94e824d
= vendor/github.com/d2g/dhcp4/LICENSE 0187683301a45e8ea393bb2ffd7889c8
= vendor/github.com/d2g/dhcp4client/LICENSE b278a92d2c1509760384428817710378
= vendor/github.com/daaku/go.zipexe/license 2285ef50c2d5083f5bb92af715e5a378
= vendor/github.com/davecgh/go-spew/LICENSE 818c0a1d81cfcfdb7ecd58db268bab7e
= vendor/github.com/daviddengcn/go-colortext/LICENSE a802db7d9a036fb71c12cf20966510dc
= vendor/github.com/dgrijalva/jwt-go/LICENSE 276f2f3ba3749d25f6a6f5fb852d462e
= vendor/github.com/docker/distribution/LICENSE d2794c0df5b907fdace235a619d80314
= vendor/github.com/docker/docker/LICENSE 9740d093a080530b5c5c6573df9af45a
= vendor/github.com/docker/go-connections/LICENSE 04424bc6f5a5be60691b9824d65c2ad8
= vendor/github.com/docker/go-units/LICENSE 04424bc6f5a5be60691b9824d65c2ad8
= vendor/github.com/docker/libnetwork/LICENSE d2794c0df5b907fdace235a619d80314
= vendor/github.com/docker/spdystream/LICENSE 246dc1c1661293602d98ff9113c3be48
= vendor/github.com/elazarl/goproxy/LICENSE e2e14e5f5bd856768da39707ed93cd41
= vendor/github.com/emicklei/go-restful/LICENSE 2ebc1c12a0f4eae5394522e31961e1de
= vendor/github.com/euank/go-kmsg-parser/LICENSE e3fc50a88d0a364313df4b21ef20c29e
= vendor/github.com/evanphx/json-patch/LICENSE 72c842ec53c3334a81b6d65b6f9025a3
= vendor/github.com/exponent-io/jsonpath/LICENSE 42f582355f11b1d4bc8615214b7f0c38
= vendor/github.com/fatih/camelcase/LICENSE.md 4c59b216ce25dc182cdb837e07ba07c0
= vendor/github.com/fsnotify/fsnotify/LICENSE c38914c9a7ab03bb2b96d4baaee10769
= vendor/github.com/GeertJohan/go.rice/LICENSE d5ffb28cde0ca932cbab7e1e1d52c20e
= vendor/github.com/ghodss/yaml/LICENSE 0ceb9ff3b27d3a8cf451ca3785d73c71
= vendor/github.com/globalsign/mgo/LICENSE 566e96676859b5704130b80941bc9f1f
= vendor/github.com/go-openapi/analysis/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/go-openapi/errors/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/go-openapi/jsonpointer/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/go-openapi/jsonreference/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/go-openapi/loads/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/go-openapi/runtime/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/go-openapi/spec/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/go-openapi/strfmt/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/go-openapi/swag/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/go-openapi/validate/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/go-ozzo/ozzo-validation/LICENSE da12d993f2ce14947ad6eec35520b081
= vendor/github.com/go-sql-driver/mysql/LICENSE 815ca599c9df247a0c7f619bab123dad
= vendor/github.com/godbus/dbus/LICENSE 09042bd5c6c96a2b9e45ddf1bc517eed
= vendor/github.com/gogo/protobuf/LICENSE f76ab0572aa28f537b1508f2f2bc155e
= vendor/github.com/golang/groupcache/LICENSE 19cbd64715b51267a47bf3750cc6a8a5
= vendor/github.com/golang/mock/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/golang/protobuf/LICENSE 14db3a56c3796a940ba32948a15f97d0
= vendor/github.com/google/btree/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/google/cadvisor/LICENSE e7790b946bfacb700e8a8f2baedb3205
= vendor/github.com/google/certificate-transparency-go/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/google/gofuzz/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/google/uuid/LICENSE 88073b6dd8ec00fe09da59e0b6dfded1
= vendor/github.com/googleapis/gnostic/LICENSE b1e01b26bacfc2232046c90a330332b3
= vendor/github.com/GoogleCloudPlatform/k8s-cloud-provider/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/gophercloud/gophercloud/LICENSE dd19699707373c2ca31531a659130416
= vendor/github.com/gorilla/websocket/LICENSE c007b54a1743d596f46b2748d9f8c044
= vendor/github.com/gregjones/httpcache/LICENSE.txt 3cfef421226b2dacde78a4871380ac24
= vendor/github.com/grpc-ecosystem/go-grpc-middleware/LICENSE 7ab5c73bb7e4679b16dd7c11b3559acf
= vendor/github.com/grpc-ecosystem/go-grpc-prometheus/LICENSE 7ab5c73bb7e4679b16dd7c11b3559acf
= vendor/github.com/grpc-ecosystem/grpc-gateway/LICENSE.txt c510a2a01572b82d27f28fd4d02ca318
= vendor/github.com/hashicorp/golang-lru/LICENSE f27a50d2e878867827842f2c60e30bfc
= vendor/github.com/hashicorp/hcl/LICENSE b278a92d2c1509760384428817710378
= vendor/github.com/heketi/heketi/LICENSE a58e72c3bda574189508cb90d56fa19f
= vendor/github.com/imdario/mergo/LICENSE ff13e03bb57bf9c52645f2f942afa28b
= vendor/github.com/inconshreveable/mousetrap/LICENSE b23cff9db13f093a4e6ff77105cbd8eb
= vendor/github.com/JeffAshton/win_pdh/LICENSE fadcabe0503181faf8d4a9579bed3b7f
= vendor/github.com/jmespath/go-jmespath/LICENSE 9abfa8353fce3f2cb28364e1e9016852
= vendor/github.com/jmhodges/clock/LICENSE bc91f6bbbc0460d728657b023f3f6596
= vendor/github.com/jmoiron/sqlx/LICENSE 4353e10ea313ec47121dd2c7d89ce6b9
= vendor/github.com/jonboulle/clockwork/LICENSE 136e4f49dbf29942c572a3a8f6e88a77
= vendor/github.com/json-iterator/go/LICENSE 0b69744b481d72d30dbf69f84a451cfd
= vendor/github.com/jstemmer/go-junit-report/LICENSE 5d54de929d5bc244d21d8d4d91a6ac64
= vendor/github.com/jteeuwen/go-bindata/LICENSE 8dcedca69f7a474372829521f37954b1
= vendor/github.com/kardianos/osext/LICENSE 591778525c869cdde0ab5a1bf283cd81
= vendor/github.com/karrick/godirwalk/LICENSE 7bea66fc0a31c6329f9392034bee75d2
= vendor/github.com/kisielk/sqlstruct/LICENSE bc5911f8d3b8034f564b63d7ea436519
= vendor/github.com/kr/fs/LICENSE 591778525c869cdde0ab5a1bf283cd81
= vendor/github.com/lib/pq/LICENSE.md 0eab29964025b358179aa6d8f7db14bf
= vendor/github.com/libopenstorage/openstorage/LICENSE 40c3e1c9eacda859a17048003909a2f8
= vendor/github.com/liggitt/tabwriter/LICENSE 5d4950ecb7b26d2c5e4e7b4e0dd74707
= vendor/github.com/lithammer/dedent/LICENSE 2f1370f16b018459285c6972a934ba0e
= vendor/github.com/magiconair/properties/LICENSE c383a608fb9a0d227953e928803b9631
= vendor/github.com/mailru/easyjson/LICENSE 819e81c2ec13e1bbc47dc5e90bb4d88b
= vendor/github.com/MakeNowJust/heredoc/LICENSE 59c4411f6d7dfdaa85623e672d3d4438
= vendor/github.com/marstr/guid/LICENSE.txt 356484d12e6ad8a7c2d360b236e9a9c8
= vendor/github.com/mattn/go-shellwords/LICENSE e5116fc268f5118168ff06a271b50ef9
= vendor/github.com/mattn/go-sqlite3/LICENSE 2b7590a6661bc1940f50329c495898c6
= vendor/github.com/matttproud/golang_protobuf_extensions/LICENSE e3fc50a88d0a364313df4b21ef20c29e
= vendor/github.com/mesos/mesos-go/LICENSE 6c4db32a2fa8717faffa1d4f10136f47
= vendor/github.com/mholt/caddy/LICENSE.txt e3fc50a88d0a364313df4b21ef20c29e
= vendor/github.com/Microsoft/go-winio/LICENSE 69205ff73858f2c22b2ca135b557e8ef
= vendor/github.com/Microsoft/hcsshim/LICENSE d4c2cbbea5ee1e7c86dff68a7073718e
= vendor/github.com/miekg/dns/LICENSE 147353de6868a20caa562d26eab7b3c5
= vendor/github.com/mindprince/gonvml/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/mistifyio/go-zfs/LICENSE cce9462224bfb44c1866ef7bd5eddf54
= vendor/github.com/mitchellh/go-wordwrap/LICENSE.md 56da355a12d4821cda57b8f23ec34bc4
= vendor/github.com/mitchellh/mapstructure/LICENSE 3f7765c3d4f58e1f84c4313cecf0f5bd
= vendor/github.com/modern-go/concurrent/LICENSE 86d3f3a95c324c9479bd8986968f4327
= vendor/github.com/modern-go/reflect2/LICENSE 86d3f3a95c324c9479bd8986968f4327
= vendor/github.com/mohae/deepcopy/LICENSE 268dc9c546e3de67a93c1d12a039d702
= vendor/github.com/mrunalp/fileutils/LICENSE 435b266b3899aa8a959f17d41c56def8
= vendor/github.com/munnerz/goautoneg/README.txt a33eda65f1bc658f358e1d690c6a93d4
= vendor/github.com/mvdan/xurls/LICENSE d459301ff3fc917904837249508d99af
= vendor/github.com/mxk/go-flowrate/LICENSE 781f3c14fa867bae781c9e409831f954
= vendor/github.com/Nvveen/Gotty/LICENSE ab11220c6af5f3758ccae7bf525f8dec
= vendor/github.com/NYTimes/gziphandler/LICENSE.md e30b94cbe70132b181f72f953fbb3c82
= vendor/github.com/onsi/ginkgo/LICENSE 570603114d52313cb86c0206401c9af7
= vendor/github.com/onsi/gomega/LICENSE 570603114d52313cb86c0206401c9af7
= vendor/github.com/opencontainers/go-digest/LICENSE.code 9cd86830b557232ce55e2a6b47387471
= vendor/github.com/opencontainers/image-spec/LICENSE 27ef03aa2da6e424307f102e8b42621d
= vendor/github.com/opencontainers/runc/LICENSE 435b266b3899aa8a959f17d41c56def8
= vendor/github.com/opencontainers/runtime-spec/LICENSE b355a61a394a504dacde901c958f662c
= vendor/github.com/opencontainers/selinux/LICENSE e3fc50a88d0a364313df4b21ef20c29e
= vendor/github.com/pborman/uuid/LICENSE 88073b6dd8ec00fe09da59e0b6dfded1
= vendor/github.com/pelletier/go-toml/LICENSE dc9ea87a81f62b8871b2a4158edbfde6
= vendor/github.com/peterbourgon/diskv/LICENSE f9f3e815fc84aa04c4f4db33c553eef9
= vendor/github.com/pkg/errors/LICENSE 6fe682a02df52c6653f33bd0f7126b5a
= vendor/github.com/pkg/sftp/LICENSE 452fc5cc5a9127a0e828d73423d45035
= vendor/github.com/pmezard/go-difflib/LICENSE e9a2ebb8de779a07500ddecca806145e
= vendor/github.com/pquerna/cachecontrol/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/pquerna/ffjson/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/prometheus/client_golang/LICENSE 86d3f3a95c324c9479bd8986968f4327
= vendor/github.com/prometheus/client_model/LICENSE 86d3f3a95c324c9479bd8986968f4327
= vendor/github.com/prometheus/common/LICENSE 86d3f3a95c324c9479bd8986968f4327
= vendor/github.com/prometheus/procfs/LICENSE 86d3f3a95c324c9479bd8986968f4327
= vendor/github.com/PuerkitoBio/purell/LICENSE fb8b39492731abb9a3d68575f3eedbfa
= vendor/github.com/PuerkitoBio/urlesc/LICENSE 591778525c869cdde0ab5a1bf283cd81
= vendor/github.com/quobyte/api/LICENSE beacc5ea3bcda24bdcec545022dbb0b4
= vendor/github.com/rancher/go-rancher/LICENSE 2ee41112a44fe7014dce33e26468ba93
= vendor/github.com/Rican7/retry/LICENSE 6421d9994f58f630cd4dd33e88991ccc
= vendor/github.com/robfig/cron/LICENSE d924a63cb54a2a6c4bd28c50b2b0af59
= vendor/github.com/rubiojr/go-vhd/LICENSE 9ce5db55ba47444787183e59733e1977
= vendor/github.com/russross/blackfriday/LICENSE.txt ecf8a8a60560c35a862a4a545f2db1b3
= vendor/github.com/satori/go.uuid/LICENSE ae4ba217c6e20c2d8f48f69966b9121b
= vendor/github.com/seccomp/libseccomp-golang/LICENSE 343b433e752e8b44a543cdf61f14b628
= vendor/github.com/shurcooL/sanitized_anchor_name/LICENSE bee2c3aa5bd0f265ffbd193eb18ca30d
= vendor/github.com/sigma/go-inotify/LICENSE 5d4950ecb7b26d2c5e4e7b4e0dd74707
= vendor/github.com/sirupsen/logrus/LICENSE 8dadfef729c08ec4e631c4f6fc5d43a0
= vendor/github.com/soheilhy/cmux/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/spf13/afero/LICENSE.txt 920d76114a32b0fb75b3f2718c5a91be
= vendor/github.com/spf13/cast/LICENSE 67fac7567cbf6ba946e5576d590b1ed4
= vendor/github.com/spf13/cobra/LICENSE.txt 920d76114a32b0fb75b3f2718c5a91be
= vendor/github.com/spf13/jwalterweatherman/LICENSE 67fac7567cbf6ba946e5576d590b1ed4
= vendor/github.com/spf13/pflag/LICENSE 1e8b7dc8b906737639131047a590f21d
= vendor/github.com/spf13/viper/LICENSE 67fac7567cbf6ba946e5576d590b1ed4
= vendor/github.com/storageos/go-api/LICENCE 10d8703157b5fd9422b1813bae555905
= vendor/github.com/stretchr/objx/LICENSE.md 21e79cfe7201b9b64535bfae0895795b
= vendor/github.com/stretchr/testify/LICENSE 39cd1d751bc25944831de86496e3cf68
= vendor/github.com/syndtr/gocapability/LICENSE a7304f5073e7be4ba7bffabbf9f2bbca
= vendor/github.com/tmc/grpc-websocket-proxy/LICENSE 0894f9b225c28f0896b4bab675a2e19a
= vendor/github.com/ugorji/go/LICENSE a03c0693c900925da5789db4e72da142
= vendor/github.com/vishvananda/netlink/LICENSE 2ade771c7d7211af507864e8dd520529
= vendor/github.com/vishvananda/netns/LICENSE 2ade771c7d7211af507864e8dd520529
= vendor/github.com/vmware/govmomi/LICENSE.txt 3b83ef96387f14655fc854ddc3c6bd57
= vendor/github.com/vmware/photon-controller-go-sdk/LICENSE 0de60303c844eac44e45012dac1987de
= vendor/github.com/xanzy/go-cloudstack/LICENSE 136e4f49dbf29942c572a3a8f6e88a77
= vendor/github.com/xiang90/probing/LICENSE 61b57bfd44f7924c8a12cd72d3aa1e93
= vendor/go.uber.org/atomic/LICENSE.txt 1caee86519456feda989f8a838102b50
= vendor/go.uber.org/multierr/LICENSE.txt f65b21a547112d1bc7b11b90f9b31997
= vendor/go.uber.org/zap/LICENSE.txt 5e8153e456a82529ea845e0d511abb69
= vendor/golang.org/x/crypto/LICENSE 5d4950ecb7b26d2c5e4e7b4e0dd74707
= vendor/golang.org/x/lint/LICENSE 4c728948788b1a02f33ae4e906546eef
= vendor/golang.org/x/net/LICENSE 5d4950ecb7b26d2c5e4e7b4e0dd74707
= vendor/golang.org/x/oauth2/LICENSE 704b1e0c436dbf193e7dcbd4cf06ec81
= vendor/golang.org/x/sys/LICENSE 5d4950ecb7b26d2c5e4e7b4e0dd74707
= vendor/golang.org/x/text/LICENSE 5d4950ecb7b26d2c5e4e7b4e0dd74707
= vendor/golang.org/x/time/LICENSE 5d4950ecb7b26d2c5e4e7b4e0dd74707
= vendor/golang.org/x/tools/LICENSE 5d4950ecb7b26d2c5e4e7b4e0dd74707
= vendor/gonum.org/v1/gonum/LICENSE 665e67d07d85e236cceb8de602c6255a
= vendor/google.golang.org/api/LICENSE a651bb3d8b1c412632e28823bb432b40
= vendor/google.golang.org/appengine/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/google.golang.org/genproto/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/google.golang.org/grpc/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/gopkg.in/gcfg.v1/LICENSE 13cea479df204c85485b5db6eb1bc9d5
= vendor/gopkg.in/inf.v0/LICENSE 13cea479df204c85485b5db6eb1bc9d5
= vendor/gopkg.in/natefinch/lumberjack.v2/LICENSE 574cdb55b81249478f5af5f789e9e29f
= vendor/gopkg.in/square/go-jose.v2/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/gopkg.in/warnings.v0/LICENSE c6775875c9d604beb22447dfae3d7049
= vendor/gopkg.in/yaml.v2/LICENSE e3fc50a88d0a364313df4b21ef20c29e
= vendor/k8s.io/gengo/LICENSE ad09685d909e7a9f763d2bb62d4bd6fb
= vendor/k8s.io/heapster/LICENSE 136e4f49dbf29942c572a3a8f6e88a77
= vendor/k8s.io/klog/LICENSE 19cbd64715b51267a47bf3750cc6a8a5
= vendor/k8s.io/kube-openapi/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/k8s.io/repo-infra/LICENSE e3fc50a88d0a364313df4b21ef20c29e
= vendor/k8s.io/utils/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
= vendor/sigs.k8s.io/kustomize/LICENSE e3fc50a88d0a364313df4b21ef20c29e
= vendor/sigs.k8s.io/structured-merge-diff/LICENSE e3fc50a88d0a364313df4b21ef20c29e
= vendor/sigs.k8s.io/yaml/LICENSE 0ceb9ff3b27d3a8cf451ca3785d73c71
= vendor/vbom.ml/util/LICENSE 9f7e1d7e8f527330ebb5f4c32e0f3e40
```
</details>

Diff between master and this PR:
```
$ diff <(git show master:Godeps/LICENSES | egrep '= vendor/.*' | grep -v "licensed under" | sort -uf) <(git show HEAD:Godeps/LICENSES | egrep '= vendor/.*' | grep -v "licensed under" | sort -uf)
54d53
< = vendor/github.com/go-ini/ini/LICENSE 19cbd64715b51267a47bf3750cc6a8a5
102,103d100
< = vendor/github.com/kr/pretty/License 9d305c2010c6891ee4f3cd42a562f78f
< = vendor/github.com/kr/text/License 449bfedd81a372635934cf9ce004c0cf
196a194
> = vendor/google.golang.org/appengine/LICENSE 3b83ef96387f14655fc854ddc3c6bd57
```

Explanation for differences:
* the go-ini dependency was unused and dropped in this PR
* the github.com/kr/... dependencies were dropped when we dropped our Godep fork
* the google.golang.org/appengine dependency was added because go modules vendor dependencies for all build tags

All other license files are still included

dims commented:
/lgtm

Please remove hold when appropriate

dims commented:
/test pull-kubernetes-kubemark-e2e-gce-big

thockin commented:
I can ACK the licenses here.  It's not the final structure I want but it is enough to build on.  We can tackle license gather separately.

liggitt commented:
fixed shellcheck error in pin-dependency.sh, CI should be green now

dims commented:
/lgtm

liggitt commented:
holding for a successful run of https://github.com/kubernetes/publishing-bot/pull/169 (publishes to https://github.com/kubernetes-nightly/client-go, etc)

liggitt commented:
/retest

liggitt commented:
draft of email to kubernetes-dev in progress at https://docs.google.com/document/d/1YCo5KLubEK4S8jGAe_LvTq9FjuUu8gxe9EnJKcGmrdM/edit (doc readable by kubernetes-dev@googlegroups.com)

dims commented:
/lgtm

sttts commented:
Rebase.

liggitt commented:
rebased

dims commented:
/lgtm

liggitt committed:
Update publishing bot smoke tests for go modules

liggitt commented:
added "Update publishing bot smoke tests for go modules" commit pulling in changes tested in https://github.com/kubernetes/publishing-bot/pull/169

dims commented:
/lgtm

liggitt commented:
/retest

liggitt commented:
/retest

liggitt commented:
publishing bot run was successful (see https://github.com/kubernetes-nightly/client-go/blob/master/go.mod for example output)

/hold cancel

fejta commented:
When will the go.mod files be live on things like k8s.io/api?

liggitt commented:
The publishing bot changes merged, a manual test run worked as expected (pushed to the kubernetes-nightly org). We're letting an automated test run pass tonight, then will push the new version of the bot live tomorrow morning. 

staging repos should have go.mod files on master by EOD tomorrow.
